contrib/valyala/fasthttp.v1/fasthttp_test.go

```diff
@@ -0,0 +1,270 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package fasthttp
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/valyala/fasthttp"
+	httptrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/net/http"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/mocktracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+const errMsg = "This is an error!"
+
+func ignoreResources(fctx *fasthttp.RequestCtx) bool {
+	return strings.HasPrefix(string(fctx.URI().Path()), "/any")
+}
+
+func getFreePort(t *testing.T) int {
+	li, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+	addr := li.Addr()
+	err = li.Close()
+	require.NoError(t, err)
+	return addr.(*net.TCPAddr).Port
+}
+
+func startServer(t *testing.T, opts ...Option) string {
+	router := func(fctx *fasthttp.RequestCtx) {
+		switch string(fctx.Path()) {
+		case "/any":
+			WrapHandler(func(c *fasthttp.RequestCtx) {
+				fmt.Fprintf(c, "Hi there!")
+			}, opts...)(fctx)
+		case "/err":
+			WrapHandler(func(c *fasthttp.RequestCtx) {
+				c.Error(errMsg, 500)
+			}, opts...)(fctx)
+		case "/customErr":
+			WrapHandler(func(c *fasthttp.RequestCtx) {
+				c.Error(errMsg, 600)
+			}, opts...)(fctx)
+		case "/contextExtract":
+			WrapHandler(func(c *fasthttp.RequestCtx) {
+				_, ok := tracer.SpanFromContext(c)
+				if !ok {
+					c.Error("No span in the request context", 500)
+				} else {
+					c.SetStatusCode(200)
+				}
+				fmt.Fprintf(c, "Hi there! RequestURI is %q", c.RequestURI())
+			}, opts...)(fctx)
+		default:
+			fctx.Error("not found", fasthttp.StatusNotFound)
+		}
+	}
```

Comment:
I think this code could be simplified to run `WrapHandler` only once:

```suggestion
router := WrapHandler(func(fctx *fasthttp.RequestCtx) {
		switch string(fctx.Path()) {
		case "/any":
			fmt.Fprintf(fctx, "Hi there!")
			return
		case "/err":
			fctx.Error(errMsg, 500)
			return
		case "/customErr":
			fctx.Error(errMsg, 600)
			return
		case "/contextExtract":
			_, ok := tracer.SpanFromContext(fctx)
			if !ok {
				fctx.Error("No span in the request context", 500)
				return
			}
			fctx.SetStatusCode(200)
			fmt.Fprintf(fctx, "Hi there! RequestURI is %q", fctx.RequestURI())
			return
		default:
			fctx.Error("not found", fasthttp.StatusNotFound)
			return
		}
	})
```

Reply:
I'm making this change manually rather than committing b/c I want to make sure I include the `opts...` arg.